docs/src/proposals/interest-bearing-tokens.md

```diff
@@ -0,0 +1,284 @@
+A total amount of bytes? Regarding security, this breaks the previous
+check that a CPI must use a subset of accounts provided to the calling program.
+
+#### Dynamic sysvars
+
+The i-token program also requires the use of dynamic sysvars. For example,
+during the `GetHoldingBalance` instruction, the program needs to know the current
+time in order to properly convert from token amount to share amount.
+
+Without dynamic sysvars, we need to create a system to tell what sysvars are
+needed for different instructions on each program, which creates complications
+for all users.
+
+### RPC
+
+The validator's RPC server needs to properly handle new token program accounts
+for read-only instructions and pre / post balances.
+
+#### `getBalance` and `getSupply`
+
+Instead of deserializing SPL token accounts and returning the data, the
+RPC server runs read-only transactions on the i-token program.
+
+RPC simply uses the same flow as on-chain programs: create an ephemeral account
+then runs the read-only transaction onto the bank requested.
+
+TODO how do we set the cost of read-only transactions? Do we lower it specially
+for read-only RPC calls? The i-token program will probably be one of the more
+cost-intensive computations possible since it is performing present value
+discounting.
+
+#### `preTokenBalances` and `postTokenBalances`
+
+As mentioned earlier in the Token Program Conformance section, the i-token
+program caches balances in the `amount` field of the SPL token holding. This
+should only be used by RPC, and not by programs to get the balance of a non-SPL
+holding.
+
+### Associated Token Program
+
+The Associated Token Program needs to support all token programs seamlessly.
+Currently, it performs the following sequence:
+
+* `transfer` required lamports
+* `allocate` space
+* `assign` to the SPL token program
+* call `InitializeAccount` on SPL token
+
+This does not work for an opaque token program, because we do not know the size required
+from the outside. Conversely, if we allow a token program to take the lamports it wants,
+a malicious token program could take too much from users.
+
+To get around this, the Token Program interface has a new `CreateHolding`
+instruction, which only allocates space and assigns the account to the program.
+Once the space is allocated, the Associated Token Program can transfer the
+required lamports, and return an error if that number is too large. The process
+becomes:
+
+* call `CreateHolding` on the token program (allocate and assign)
+* calculate rent requirement based on the data size of holding account
+* `transfer` required lamports
+* call `InitializeAccount` on the token program
+
+### Web3 / Wallets
+
+To properly support i-tokens, wallets and applications will need to make the
+following changes:
+
+* get holdings: query for holdings owned by the public key, for all official
+token programs listed in the registry (not just `Tokenkeg...`)
```

Comment:
this means `getTokenAccountsByOwner` needs smarter secondary indexes.

Reply:
Very true, I'll add that as a consideration